Heuristics.php

```diff
@@ -0,0 +1,164 @@
+        if ($reference !== '') $reference = $this->checkVariable($reference);
+        if ($reference !== '') $reference = $this->checkFunction($reference);
+        if ($reference !== '') $reference = $this->checkPSRClass($reference);
+        if ($reference !== '') $this->def = $reference;
+    }
+
+    /**
+     * @return string
+     */
+    public function getDef()
+    {
+        return trim(preg_replace('/[^\w]+/', '', $this->def));
+    }
+
+    /**
+     * @return string
+     */
+    public function getPath()
+    {
+        return trim(preg_replace('/[^\w.]+/', ' ', $this->path));
+    }
+
+    /**
+     * Handle things in the form path#symbol
+     *
+     * @param string $reference
+     * @return string
+     */
+    protected function checkHash($reference)
+    {
+        if (strpos($reference, '#') === false) return $reference;
+        list($this->path, $this->def) = explode('#', $reference, 2);
+        return '';
+    }
+
+    /**
+     * Known file extension?
+     *
+     * @param string $reference
+     * @return mixed|string
+     */
+    protected function checkFilename($reference)
+    {
+        if (preg_match('/\.(php|js|css|html)$/', $reference)) {
+            $this->def = '';
+            $this->path = $reference;
+            return '';
+        }
+        return $reference;
+    }
+
+    /**
+     * Namespaces are paths
+     *
+     * @param string $reference
+     * @return string
+     */
+    protected function checkNamespace($reference)
+    {
+        if (strpos($reference, '\\') === false) return $reference;
+
+        $parts = explode('\\', $reference);
+        $parts = array_filter($parts);
+        $reference = array_pop($parts); // last part may be more than a class
+
+        // our classes are in inc
+        if($parts[0] == 'dokuwiki') $parts[0] = 'inc';
+
+        $this->path = join(' ', $parts);
+
+        return $reference;
+    }
+
+    /**
+     * Is there something called on a class?
+     */
+    protected function checkClassPrefix($reference)
+    {
+        if (
+            strpos($reference, '::') === false &&
+            strpos($reference, '->') === false
+        ) {
+            return $reference;
+        }
+        list($class, $reference) = preg_split('/(::|->)/', $reference, 2);
+
+        $this->path .= ' ' . $class;
+        $this->def = $reference;
+        return '';
+    }
+
+    /**
+     * Clearly a variable
+     *
+     * @param string $reference
+     * @return string
+     */
+    protected function checkVariable($reference)
+    {
+        if ($reference[0] == '$') {
+            $this->def = $reference;
+            return '';
+        }
+        return $reference;
+    }
+
+    /**
+     * It's a function
+     *
+     * @param string $reference
+     * @return string
+     */
+    protected function checkFunction($reference)
+    {
+        if (substr($reference, -2) == '()') {
+            $this->def = $reference;
+            return '';
+        }
```

Comment:
```suggestion
        }
        if (preg_match('/\(.+?\)$/', $reference)) {
            [$reference, /* $arguments */] = explode('(', $reference, 2);
            $this->def = $reference;
            return '';
        }
```
Just ignore eventually added arguments. Often examples are given with the arguments, this reduces the effort because no title is needed anymore.